lib/src/scanner.dart

```diff
@@ -2,4 +2,219 @@
+      _scanner.readChar();
+    }
+  }
+
+  String get _lastString => _scanner.lastMatch.group(0);
+}
+
+abstract class Token {
+  const Token();
+
+  @override
+  String toString() {
+    throw '$runtimeType forgot to implement toString()';
+  }
+}
+
+class Identifier extends Token {
+  const Identifier(this.name);
+
+  final String name;
+
+  @override
+  String toString() => '${runtimeType}(${name})';
+}
+
+class Keyword extends Token {
+  static const $import = const Keyword('import');
+  static const $widget = const Keyword('widget');
+  static const $state = const Keyword('state');
+  static const $build = const Keyword('build');
+
+  static const List<Keyword> values = const <Keyword>[
+    $import,
+    $widget,
+    $state,
+    $build,
+  ];
+
+  static final List<String> names = new List.unmodifiable(values.map((k) => k.name));
+
+  const Keyword(this.name);
+
+  static final Map<String, Keyword> _nameToKeyword = new Map<String, Keyword>.fromIterable(
+    Keyword.values,
+    key: (k) => k.name,
+  );
+
+  static final Map<Keyword, String> _keywordToName = new Map<Keyword, String>.fromIterable(
+    Keyword.values,
+    value: (k) => k.name,
+  );
+
+  static Keyword lookupByName(String name) {
+    assert(name != null);
+    return _nameToKeyword[name];
+  }
+
+  final String name;
+
+  @override
+  String toString() => '${runtimeType}("${name}")';
+}
+
+class Syntax extends Token {
+  static const $openBrace = const Syntax('{', r'\{');
+  static const $closeBrace = const Syntax('}', r'\}');
+  static const $openParen = const Syntax('(', r'\(');
+  static const $closeParen = const Syntax(')', r'\)');
+  static const $lt = const Syntax('<', r'<');
+  static const $gt = const Syntax('>', r'>');
+  static const $dot = const Syntax('.', r'\.');
+  static const $comma = const Syntax(',', r'\,');
+  static const $equals = const Syntax('==', r'==');
+  static const $assignment = const Syntax('=', r'=');
+  static const $questionMark = const Syntax('?', r'\?');
+  static const $exclamationMark = const Syntax('!', r'\!');
+  static const $dollar = const Syntax('!', r'\$');
+  static const $colon = const Syntax(':', r'\:');
+  static const $slash = const Syntax('/', r'/');
+
+  static const List<Syntax> values = const <Syntax>[
+    $openBrace,
+    $closeBrace,
+    $openParen,
+    $closeParen,
+    $lt,
+    $gt,
+    $dot,
+    $comma,
+    $equals,
+    $assignment,
+    $questionMark,
+    $exclamationMark,
+    $dollar,
+    $colon,
+    $slash,
+  ];
+
+  static final Map<String, Syntax> _nameToSyntax = new Map<String, Syntax>.unmodifiable(
+    new Map<String, Syntax>.fromIterable(
+      values,
+      key: (s) => s.token,
+    )
+  );
+
+  static Syntax lookupByName(String name) => _nameToSyntax[name];
+
+  const Syntax(this.token, this.pattern);
+
+  final String token;
+  final String pattern;
+
+  @override
+  String toString() => '${runtimeType}("${token}")';
+}
+
+class _Patterns {
```

Comment:
It's probably fine to start with `RegExp`s but I would think about codegen instead (in the future).

Reply:
Agreed.